File: Medium/54 - SpiralMatrix/solution.ts
/**
 * Question 54. Spiral Matrix (https://leetcode-cn.com/problems/spiral-matrix/)
 *
 * 执行用时：84 ms, 在所有 typescript 提交中击败了 20.69% 的用户
 * 内存消耗：37.2 MB, 在所有 typescript 提交中击败了 100.00% 的用户
 *
 * @param {number[][]} matrix
 * @returns {number[]}
 */
function spiralOrder(matrix: number[][]): number[] {
  const m: number = matrix.length;
  if (m === 0) {
    return [];
  }

  const n: number = matrix[0].length;
  if (n === 0) {
    return [];
  }

  let rowStart: number = 0;
  let rowEnd: number = m - 1;
  let columnStart: number = 0;
  let columnEnd: number = n - 1;

  let currentRow: number = rowStart;
  let currentColumn: number = columnStart;

  let horizontal: boolean = true;
  let horizontalDirection: number = 1;
  let verticalDirection: number = 1;

  const matrixSize: number = m * n;

  const visitedCells: number[] = [];

  while (visitedCells.length !== matrixSize) {
    visitedCells.push(matrix[currentRow][currentColumn]);

    if (horizontal) {
      if (horizontalDirection > 0) {
        // top right cornor
        if (currentColumn === columnEnd) {
          horizontal = !horizontal;
          currentRow += verticalDirection;
          horizontalDirection *= -1;
          rowStart++;
        } else {
          currentColumn++;
        }
      } else {
        // bottom left cornor
        if (currentColumn === columnStart) {
          horizontal = !horizontal;
          currentRow += verticalDirection;
          horizontalDirection *= -1;
          rowEnd--;
        } else {
          currentColumn--;
        }
      }
    } else {
      if (verticalDirection > 0) {
        // bottom right cornor
        if (currentRow === rowEnd) {
          horizontal = !horizontal;
          currentColumn += horizontalDirection;
          verticalDirection *= -1;
          columnEnd--;
        } else {
          currentRow++;
        }
      } else {
        // top left cornor
        if (currentRow === rowStart) {
          horizontal = !horizontal;
          currentColumn += horizontalDirection;
          verticalDirection *= -1;
          rowStart++;
        } else {
          currentRow--;
        }
      }
    }
  }

  return visitedCells;
}

File: Medium/54 - SpiralMatrix/Solution.cs
// Question 54. Spiral Matrix (https://leetcode-cn.com/problems/spiral-matrix/)

/// <summary>
/// 执行用时：264 ms, 在所有 C# 提交中击败了 97.71% 的用户
/// 内存消耗：29.7 MB, 在所有 C# 提交中击败了 34.31% 的用户
/// 
/// Acknowledgement: This solution borrows some idea from the 
/// [official solution](https://leetcode-cn.com/problems/spiral-matrix/solution/luo-xuan-ju-zhen-by-leetcode-solution/)
/// </summary>
public class Solution 
{
    public IList<int> SpiralOrder(int[][] matrix) 
    {
        int m = matrix.Count();
        int n = matrix[0].Count();

        IList<int> elementsInSpiral = new List<int>();
        
        int left = 0;
        int right = n - 1;
        int top = 0;
        int bottom = m - 1;

        while (top <= bottom && left <= right)
        {
            // upper bound.
            for (int j = left; j <= right; j++) {
                int element = matrix[top][j];
                elementsInSpiral.Add(element);
            }

            // right bound.
            for (int i = top + 1; i <= bottom; i++) {
                int element = matrix[i][right];
                elementsInSpiral.Add(element);
            }

            if (left < right && top < bottom) 
            {
                // lower bound.
                for (int j = right - 1; j > left; j--) {
                    int element = matrix[bottom][j];
                    elementsInSpiral.Add(element);
                }

                // left bound.
                for (int i = bottom; i > top; i--) {
                    int element = matrix[i][left];
                    elementsInSpiral.Add(element);
                }
            }

            top++;
            right--;
            bottom--;
            left++;

            System.Console.WriteLine($"top = {top}, bottom = {bottom}, left = {left}, right = {right}");
            System.Console.WriteLine($"elementsInSpiral = [{string.Join(", ", elementsInSpiral)}]");
        }

        return elementsInSpiral;
    }
}
